ClientRuntimes/CSharp/ClientRuntime.Azure/Authentication/ActiveDirectoryApplicationTokenProvider.cs

```diff
@@ -0,0 +1,153 @@
+        /// <param name="cache">The TokenCache to use during authentication.</param>
+        internal ActiveDirectoryApplicationTokenProvider(string domain, string clientId, string secret, AzureEnvironment environment, TokenCache cache)
+        {
+             if (string.IsNullOrWhiteSpace(secret))
+            {
+                throw new ArgumentOutOfRangeException("secret");
+            }
+
+            InitializeAuthenticationContext(domain, clientId, environment, cache);
+            this._credential = new ClientCredential(clientId, secret);
+            ValidateAuthenticationResult(Authenticate().Result);
+       }
+
+        /// <summary>
+        /// Returns the token type of the returned token.
+        /// </summary>
+        public string TokenType
+        {
+            get { return _type; }
+        }
+
+        /// <summary>
+        /// Gets an access token from the token cache or from AD authentication endpoint. 
+        /// Attempts to refresh the access token if it has expired.
+        /// </summary>
+        public virtual async Task<string> GetAccessTokenAsync(CancellationToken cancellationToken)
+        {
+            var result = await this.Authenticate().ConfigureAwait(false);
+            ValidateAuthenticationResult(result);
+            return result.AccessToken;
+        }
+
+        /// <summary>
+        /// Initialize private fields in the token provider.
+        /// </summary>
+        /// <param name="domain">The domain or tenant id for the application.</param>
+        /// <param name="clientId">The client Id of the application in Active Directory.</param>
+        /// <param name="environment">The Azure environment to manage resources in.</param>
+        /// <param name="cache">The TokenCache to use while authenticating.</param>
+        protected void InitializeAuthenticationContext(string domain, string clientId,
+            AzureEnvironment environment, TokenCache cache = null)
+        {
+            ValidateCommonParameters(clientId, domain, environment);
+            this._clientId = clientId;
+            this._tokenAudience = environment.TokenAudience.ToString();
+            try
+            {
+                this._authenticationContext = (cache == null)
+                    ? new AuthenticationContext(environment.AuthenticationEndpoint + domain,
+                        environment.ValidateAuthority)
+                    : new AuthenticationContext(environment.AuthenticationEndpoint + domain,
+                        environment.ValidateAuthority,
+                        cache);
+            }
+            catch (AdalException authenticationException)
+            {
+                throw new AuthenticationException(Resources.ErrorCreatingAuthenticationContext, authenticationException);
+            }
+        }
+        /// <summary>
+        /// Set the ActiveDirectory authentication properties for this user
```

Comment:
This comments doesn't match the method name of "Validate", should we rename the method to be "SetAuthenticationResult"?